### ISSUE TITLE
Soundsync controller does not load in Firefox 81

### ISSUE THREAD

**Issue by balaji-dutt** [geekuillaume/soundsync#42]
Hello,

I have installed Soundsync on a Ubuntu machine and am able to successfully access the controller in Chrome. However, when I launch the Controller URL in Firefox it hangs on the "Connecting" screen. Looking at the Console in Firefox shows the following error message:

```
Uncaught (in promise) TypeError: Ln.RTCPeerConnection is not a constructor
    initWebrtcIfNeeded http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    initiateConnection http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    connect http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    joinPeerWithHttpApi http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    Jr http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    n http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:51
    ni http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    ni http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    hs http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    Cc http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    unstable_runWithPriority http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:105
    Gi http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    Ec http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    oc http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    Vi http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    unstable_runWithPriority http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:105
    Gi http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    Vi http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    Yi http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    sc http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    Vc http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    render http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    <anonymous> http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    i http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:1
    <anonymous> http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:1
    <anonymous> http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:1
main-25f2404c7752762a0c44.js:130:57743
Uncaught (in promise) TypeError: Ln.RTCPeerConnection is not a constructor
    initWebrtcIfNeeded http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    initiateConnection http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    connect http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    joinPeerWithHttpApi http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    Jr http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    n http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:51
    ni http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    ni http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    hs http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    Cc http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    unstable_runWithPriority http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:105
    Gi http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    Ec http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    oc http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    Vi http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    unstable_runWithPriority http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:105
    Gi http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    Vi http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    Yi http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    sc http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    Vc http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    render http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:97
    <anonymous> http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:130
    i http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:1
    <anonymous> http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:1
    <anonymous> http://192.168.XX.XX:6512/main-25f2404c7752762a0c44.js:1
```

### Version Info
- Firefox 81.0.1
- Windows 10 (v1903)

**Comment by gingerlime** [comment]
Also seeing CORS errors on Firefox, which causes the controller page to continuously reload. Perhaps my Firefox privacy settings are a bit more strict though? (It's set to block cross-site and social media trackers)

![Screenshot 2020-10-11 at 09 18 59](https://user-images.githubusercontent.com/524274/95672663-197b0700-0ba3-11eb-865a-d2851f5da703.png)

**Comment by geekuillaume** [comment]
The CORS errors are expected, this is a necessary hack to establish a connection between a https context (soundsync.app) and a non-https endpoint (Soundsync on the local network, local IP address: no SSL certificate). In some cases, the Webrtc constructor is blocked by browser extensions. I've added a check for this which will show a human-readable error message in this case. Can you try again?

**Comment by balaji-dutt** [comment]
@geekuillaume I just downloaded v0.4.4 and have some good news

- If I enable an extension that blocks WebRTC, I get the following error message:
    > Your browser doesn't support the necessary web features to handle Soundsync. Chrome and Firefox are known to support Soundsync. Please download a compatible browser or update yours to use Soundsync.
    > WebRTC is blocked, check your extensions

- As @gingerlime points out, it is not enough to simply enable WebRTC. the Firefox "Enhanced Tracking Protections" option still causes the spinner to appear. I had to set "Enhanced Tracking Protection" to _off_ for the Controller page. Once I disabled tracking protection, the UI started working. There is still one glitch in Firefox which I will file as a separate issue.

There is one more good news with v0.4.4 as well. When I run v0.4.4 with AirConnect v0.2.27.0 (Aug  3 2020 build) I am able to successfully play music to my Sonos One! 🥳 🚀

**Comment by geekuillaume** [comment]
Good, so now we know why it wasn't working. I need to take a closer look at the Enhanced Tracking Protections from Firefox but I don't think I'll find a way to make it work with them. Connecting from a secure webpage (https) to a insecure http endpoint on the local network is still a unsolved problem. I've implemented multiple "hacks" to make it work but it's still not perfect. I'm thinking of adding a way to "synchronize" multiple Soundsync devices through a one-time token needed to be manually set in the WebUI but it's not the best UX-side.

Related to AirConnect, I need to spend more time looking at the code but I'll most likely add native UPnP integration to Soundsync (minus the precise synchronization which I don't think is possible with UPnP) to support TVs and older Sonos devices.

**Comment by balaji-dutt** [comment]
> 
> Related to AirConnect, I need to spend more time looking at the code but I'll most likely add native UPnP integration to Soundsync (minus the precise synchronization which I don't think is possible with UPnP) to support TVs and older Sonos devices.

One of the problems I'm seeing with AirConnect (which I believe also uses UPnP) is that the assigned port for a particular device is not consistent between reboots, i.e., if either the AirConnect service restarts or the Sonos device restarts when AirConnect re-connects, it assigns a new port for the same device. This causes a problem for SoundSync since the old port number is still in the config.

I'm not sure if this is a UPnP limitation or something specific to the AirConnect implementation, but thought I should mention it.

**Comment by geekuillaume** [comment]
As the original issue is fixed (Firefox not working) I'm closing this. Feel free to reopen if you still have a problem with this.

**Comment by gingerlime** [comment]
there's a different problem now... when I open the controller via the menu, it opens localhost:6512 which then redirects to the landing page rather than the controller. If I go to http://localhost:6512/controller then it just spins forever without connecting. https://soundsync.app/controller however still works

**Comment by geekuillaume** [comment]
I've changed the default URL to http://localhost:6512/controller instead of http://localhost:6512/. What browser are you using?

**Comment by gingerlime** [comment]
Firefox
